### PR TITLE
[Phi2] Add support for phi2 models

### PR DESCRIPTION
**Pull request by susnato**
# What does this PR do?

This PR fixes the conversion script for `phi` to include the changes made in this [commit](https://huggingface.co/microsoft/phi-1/commit/bbace889388b8bb0ba6ec3d28dcccca00f962062). It also adds a integration test for `phi2`. 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@gugarosa @ArthurZucker 

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Hey! Thanks for the PR, alright overall, but let's not add the phi2 changes here since it's not supported!

**Comment by susnato** [comment]
> Hey! Thanks for the PR, alright overall, but let's not add the phi2 changes here since it's not supported!

Hi @ArthurZucker, phi2 has the same modeling code as the library phi model...so I guess it's supported. 
I also created [susnato/phi-2](https://huggingface.co/susnato/phi-2) to verify the outputs and they are ok.

I am sorry if I am missing something here...

**Comment by ArthurZucker** [comment]
Oh no I checked #28163 and thought we needed a lot more changes! If this is enough let's rename this one to `[Phi2] Add support for phi2 models` and close the other one?

**Comment by susnato** [comment]
Hi @ArthurZucker, yes this is enough for support of `phi2`(weights conversion script and verify logits in test) but #28163 is for adding more features for both phi1 and phi2 models so we should not close that IMO.

**Comment by ArthurZucker** [comment]
Alright! Got it!

File: src/transformers/models/phi/convert_phi_weights_to_hf.py [hunk 1]
```diff
@@ -104,27 +100,39 @@ def _download(url: str, root: str):
...
-        model_path = os.path.join(checkpoint_path, each_model_name + "_" + each_model_url.split("/")[-1])
-        if not os.path.exists(model_path):
-            print(f"\n{each_model_name} was not found! Downloading it to {model_path}")
-            _download(url=each_model_url, root=model_path)
-        model_checkpoint = torch.load(model_path, map_location=device)
-        model_type = each_model_name.split("/")[1]  # phi-1 or phi-1_5
-        config = PhiConfig.from_pretrained(f"susnato/{model_type}_dev")
+        model_checkpoint = {}
+
+        # for phi-2 the weights are stored in 2 different safetensors file so we need to iterate over thaat list and download one at a time
```
Comment (ArthurZucker) [inline src/transformers/models/phi/convert_phi_weights_to_hf.py:112]:
```suggestion
        # for phi-2 the weights are stored in 2 different safetensors file so we need to iterate over that list and download one at a time
```

File: src/transformers/models/phi/convert_phi_weights_to_hf.py [hunk 2]
```diff
@@ -104,27 +100,39 @@ def _download(url: str, root: str):
...
+            if model_path.endswith("safetensors"):
+                loaded_weights = safetensors.torch.load_file(model_path, device=device)
+            else:
+                loaded_weights = torch.load(model_path, map_location=device)
+            model_checkpoint.update(**loaded_weights)
+
+        model_type = model_name.split("/")[1]  # phi-1 or phi-1_5 or phi-2
+        config = PhiConfig.from_pretrained(
+            f"susnato/{model_type}_dev" if model_name != "microsoft/phi-2" else f"susnato/{model_type}"
+        )
```
Comment (ArthurZucker) [inline src/transformers/models/phi/convert_phi_weights_to_hf.py:128]: Would rather have something explicit for the config, with a mapping to not rely on external repo!

File: src/transformers/models/phi/convert_phi_weights_to_hf.py [hunk 3]
```diff
@@ -172,4 +186,11 @@ def convert_phi_weights(checkpoint_path, pytorch_dump_folder_path, use_cuda, sav
...
 
     args = parser.parse_args()
-    convert_phi_weights(args.checkpoint_path, args.pytorch_dump_folder_path, args.use_cuda, args.save_weights_directly)
+    convert_phi_weights(
+        args.model_name,
+        args.checkpoint_path,
+        args.pytorch_dump_folder_path,
+        args.use_cuda,
+        args.save_weights_directly,
+        _MODELS,
```
Comment (ArthurZucker) [inline src/transformers/models/phi/convert_phi_weights_to_hf.py:195]: should not need to be passed as it's a global mapping

File: tests/models/phi/test_modeling_phi.py [hunk 4]
```diff
@@ -432,3 +432,19 @@ def test_model_phi_1_5_logits(self):
         EXPECTED_OUTPUT = torch.tensor([[12.2922, 13.3507,  8.6963,  9.1355,  9.3502,  9.2667, 14.2027, 13.1363, 13.5446, 11.1337,  9.9279, 16.7195, 13.0768, 14.9141, 11.9965,  8.0233, 10.3129, 10.6118, 10.0204,  9.3827,  8.8344,  8.2806,  8.0153,  8.0540, 7.0964, 16.5743, 11.1256,  9.6987, 11.4770, 10.5440], [12.3323, 14.6050,  8.9986,  8.1580,  9.5654,  6.6728, 12.5966, 12.6662, 12.2784, 11.7522,  8.2039, 16.3102, 11.2203, 13.6088, 12.0125,  9.1021, 9.8216, 10.0987,  9.0926,  8.4260,  8.8009,  7.6547,  6.8075,  7.7881, 7.4501, 15.7451, 10.5053,  8.3129, 10.0027,  9.2612]]).to(torch_device)  # fmt: skip
 
         self.assertTrue(torch.allclose(EXPECTED_OUTPUT, output[0, :2, :30], atol=1e-4, rtol=1e-4))
+
+    def test_model_phi_2_logits(self):
```
Comment (ArthurZucker) [inline tests/models/phi/test_modeling_phi.py:436]: Let's add 1 generation test as well!

**Review by ArthurZucker — COMMENTED** [review]
Looks great, a few nits left!

File: src/transformers/models/phi/convert_phi_weights_to_hf.py (hunk 3 above)
Comment (susnato) [reply src/transformers/models/phi/convert_phi_weights_to_hf.py:195]: Hmm yes but if I remove this, I am getting referenced before assignment error...

**Comment by susnato** [comment]
Hi @ArthurZucker, I have pushed the changes.

File: src/transformers/models/phi/convert_phi_weights_to_hf.py [hunk 5]
```diff
@@ -104,27 +102,48 @@ def _download(url: str, root: str):
...
+                loaded_weights = safetensors.torch.load_file(model_path, device=device)
+            else:
+                loaded_weights = torch.load(model_path, map_location=device)
+            model_checkpoint.update(**loaded_weights)
+
+        model_type = model_name.split("/")[1]  # phi-1 or phi-1_5 or phi-2
+
+        # init the config for phi-1 and phi-1.5
+        config = PhiConfig()
+        # if we are dealing with phi-2 then update the config
```
Comment (ArthurZucker) [inline src/transformers/models/phi/convert_phi_weights_to_hf.py:131]: thanks 😉

File: src/transformers/models/phi/convert_phi_weights_to_hf.py [hunk 6]
```diff
@@ -104,27 +102,48 @@ def _download(url: str, root: str):
...
+        if model_type == "phi-2":
+            config.hidden_size = 2560
+            config.intermediate_size = 10240
+            config.num_hidden_layers = 32
+            config.resid_pdrop = 0.1
+            config.partial_rotary_factor = 0.4
+            config.num_hidden_layers = 32
+            config.torch_dtype = "float16"
```
Comment (ArthurZucker) [inline src/transformers/models/phi/convert_phi_weights_to_hf.py:139]:
```suggestion
            config = PhiConfig(
                hidden_size = 2560,
                intermediate_size = 10240, 
                num_hidden_layers = 32,
                resid_pdrop = 0.1,
                partial_rotary_factor = 0.4,
                num_hidden_layers = 32,
                torch_dtype = "float16"
            )
```

File: src/transformers/models/phi/convert_phi_weights_to_hf.py [hunk 7]
```diff
@@ -104,27 +102,48 @@ def _download(url: str, root: str):
     )
 
 
-def convert_phi_weights(checkpoint_path, pytorch_dump_folder_path, use_cuda, save_weights_directly):
+def convert_phi_weights(
+    model_name, checkpoint_path, pytorch_dump_folder_path, use_cuda, save_weights_directly, _MODELS
+):
+    _MODELS = _MODELS if model_name not in _MODELS.keys() else {model_name: _MODELS.get(model_name)}
```
Comment (ArthurZucker) [inline src/transformers/models/phi/convert_phi_weights_to_hf.py:108]:
```suggestion
    model_name, checkpoint_path, pytorch_dump_folder_path, use_cuda, save_weights_directly):
    pretrained_models = _MODELS if model_name not in _MODELS.keys() else {model_name: _MODELS.get(model_name)}
```
and just use `pretrained_models`

File: src/transformers/models/phi/convert_phi_weights_to_hf.py [hunk 8]
```diff
@@ -104,27 +102,48 @@ def _download(url: str, root: str):
...
     device = "cuda" if torch.cuda.is_available() and use_cuda else "cpu"
-    for each_model_name, each_model_url in _MODELS.items():
+    for model_name, model_url in _MODELS.items():
```
Comment (ArthurZucker) [inline src/transformers/models/phi/convert_phi_weights_to_hf.py:110]:
```suggestion
    for model_name, model_url in pretrained_models.items():
```

**Review by ArthurZucker — APPROVED** [review]
LGTM thanks @susnato for the continued efforts 🤗

**Comment by ArthurZucker** [comment]
Btw it’s mostly nits we can merge without them 🤗

**Comment by susnato** [comment]
Hi @ArthurZucker, sorry I forgot to update the doc file to include the `phi2` example, I guess we should add an example and show it because many people were unable to use it with the library and raised issues in `microsoft/phi-2` (until the weights are in appropriate order we should use `susnato/phi-2` instead of the official repo to load it successfully). If you agree, I would create a quick PR to add the example. WDYT?

**Comment by ArthurZucker** [comment]
Yes that would be great. We'll do a release this week as well so that everything will be included! 🤗
